lib/membrane_rtc_engine/endpoints/webrtc/track_sender.ex

```diff
@@ -0,0 +1,51 @@
+defmodule Membrane.RTC.Engine.Endpoint.WebRTC.TrackSender do
+  @moduledoc false
+
+  # TrackSender:
+  # * adds `is_keyframe` flag to each buffer's metadata
+  # (will be removed after releasing new RTP plugin)
+  # * tracks encoding activity
+
+  use Membrane.Filter
+
+  require Membrane.Logger
+
+  alias Membrane.RTC.Engine.Endpoint.WebRTC.TrackSenderState
```

Comment:
With the current implementation, it seems this shouldn't be a separate module (besides the struct that could be just nested here), but I'm guessing the goal is to separate the element from the logic of tracking encodings and so on. In such a case, I believe the naming should be different - imagine `TrackSenderState` as something that could be used as a standalone entity, like a `MapSet`.

I'm not sure what kind of logic will it contain (right now I don't see any encoding activity tracking) but I'm thinking of a name like `(Outbound)TrackTracker`, maybe you'll come up with something better

Reply:
The idea here is that we have a state and a bunch of functions that modify this state and return some actions.

This makes testing much more easier as I don't have to create pipelines just for sending one or two buffers. I could directly call `handle_notification` but I would have to mock `ctx` and it also doesn't seem as pretty as the separate module.

Treating `TrackSenderState` as a standalone entity is also problematic as we have to generate actions that are Membrane specific and they need specific output pad.

Maybe I could rename this to: 
* `TrackSenderImpl`
* `TrackSenderCore`
* `TrackSenderEngine`
* `TrackSenderWorker`
* `TrackSenderAgent`

Reply:
> This makes testing much more easier as I don't have to create pipelines just for sending one or two buffers. 

The trouble is that it should be done anyway - we had cases in other plugins where the unit tests were passing fine, but the element was crashing when used as part of the pipeline (e.g. no caps before buffers)

> I could directly call `handle_notification` but I would have to mock `ctx` and it also doesn't seem as pretty as the separate module.
>
> Treating `TrackSenderState` as a standalone entity is also problematic as we have to generate actions that are Membrane specific and they need specific output pad.

The last sentence indicates this really shouldn't be a separate module - the coupling between functions from these two modules is tight and IMHO separating only complicates the element as you need to jump between the two modules back and forth when figuring them out.

If testing is the reason for such separation, then I'd suggest moving the functions from `State` module back to the element with `@doc false`

Reply:
> The trouble is that it should be done anyway - we had cases in other plugins where the unit tests were passing fine, but the element was crashing when used as part of the pipeline (e.g. no caps before buffers)

I will try to write some pipelines but at a first glance it sounds to me like an overhead 

> If testing is the reason for such separation, then I'd suggest moving the functions from State module back to the element with @doc false

I am not sure if I got this, we would end up with one module and code like 

```elixir
def handle_process(state, buffer) do
  process(state, buffer)
end

test "" do
  assert X.process(state, buffer) == y
end
```

Reply:
Yep, except that `handle_process` has different arguments. IMHO that's more or less what you've done, with the addition of moving `X.process` to `XState.process` and I don't see a good reason for that while generating spaghetti-like code.

The separate module would make sense if it were less coupled to the element and was an abstraction of its own with an API you could imagine using outside - like in tests. Otherwise, you should test the element's API which is callbacks